src/shr_transport.h4

```diff
@@ -0,0 +1,594 @@
+            }
+            break;
+        /* Note: The following ops are only used by AMO reductions, which are
+         * presently disabled when shared memory AMOs are enabled. */
+        case SHM_INTERNAL_PROD:
+        case SHM_INTERNAL_MIN:
+        case SHM_INTERNAL_MAX:
+        default:
+            RAISE_ERROR_MSG("Unsupported op op=%d, dtype=%d\n", op, datatype);
+    }
+
+#undef SHMEM_DEF_BAND_OP
+#undef SHMEM_DEF_BOR_OP
+#undef SHMEM_DEF_BXOR_OP
+#undef SHMEM_DEF_SUM_OP
+
+#else
+    RAISE_ERROR_STR("No path to peer");
+#endif
+}
+
+
+static inline
+void
+shmem_shr_transport_atomic_fetch(shmem_ctx_t ctx, void *target,
+                                 const void *source, size_t len,
+                                 int pe, shm_internal_datatype_t datatype)
+{
+#if USE_SHR_ATOMICS
+    int noderank = shmem_internal_get_shr_rank(pe);
+    void *remote_ptr;
+
+    if (noderank == -1)
+        RAISE_ERROR_MSG("No shared memory path to peer %d\n", pe);
+
+    shmem_shr_transport_ptr((void *)source, noderank, &remote_ptr);
+
+#define SHMEM_DEF_FETCH(STYPE,TYPE,ITYPE)                                       \
+        case ITYPE:                                                             \
+            __atomic_load((TYPE *)remote_ptr, (TYPE *)target, __ATOMIC_SEQ_CST);\
+            break;
+
+    switch (datatype) {
+SHMEM_DEFINE_FOR_EXTENDED_AMO(SHMEM_DEF_FETCH)
+        default:
+            RAISE_ERROR_MSG("Unsupported datatype dtype=%d\n", datatype);
+    }
+#undef SHMEM_DEF_FETCH
+
+#else
+    RAISE_ERROR_STR("No path to peer");
+#endif
+}
+
+
+static inline
+void
+shmem_shr_transport_atomic_set(shmem_ctx_t ctx, void *target,
+                               const void *source, size_t len,
+                               int pe, shm_internal_datatype_t datatype)
+{
+#if USE_SHR_ATOMICS
+    int noderank = shmem_internal_get_shr_rank(pe);
+    void *remote_ptr;
+
+    if (noderank == -1)
+        RAISE_ERROR_MSG("No shared memory path to peer %d\n", pe);
+
+    shmem_shr_transport_ptr(target, noderank, &remote_ptr);
+
+#define SHMEM_DEF_SET(STYPE,TYPE,ITYPE)                                                 \
+        case ITYPE:                                                                     \
+            __atomic_store((TYPE *)remote_ptr, (TYPE *)source, __ATOMIC_SEQ_CST);       \
+            break;
+
+    switch (datatype) {
+SHMEM_DEFINE_FOR_EXTENDED_AMO(SHMEM_DEF_SET)
+        default:
+            RAISE_ERROR_MSG("Unsupported datatype dtype=%d\n", datatype);
+    }
+#undef SHMEM_DEF_SET
+
+#else
+    RAISE_ERROR_STR("No path to peer");
+#endif
+}
+
+
+static inline
+void
+shmem_shr_transport_atomicv(shmem_ctx_t ctx, void *target, const void *source,
+                            size_t len, int pe, shm_internal_op_t op,
+                            shm_internal_datatype_t datatype)
+{
+#if USE_SHR_ATOMICS
+    RAISE_ERROR_STR("No path to peer");
```

Comment:
No `atomicv` I presume?

Reply:
Have you thought about how one might implement it?  It's not meant to be atomic w.r.t updating the entire vector, correct?  Could you do a relatively simple loop of individual atomic ops?

Reply:
Nice observation -- this is something that I consciously left to future work. At the moment `atomicv` usage (only used by allreduce) is disabled via `shmem_transport_atomic_supported` in the transport layer.